lib/index.js

```diff
@@ -2,115 +2,113 @@
   subject: Joi.string()
 }).label('jwt auth strategy options');
 
 // Declare internals
 var internals = {};
 
-
-exports.register = function (server, options, next) {
-
+function register(server, options) {
   server.auth.scheme('jwt', internals.implementation);
-  next();
-};
-
-exports.register.attributes = {
-    pkg: require('../package.json')
 };
 
 function isFunction(functionToCheck) {
-  return Object.prototype.toString.call(functionToCheck) === '[object Function]';
+  const objectProto = Object.prototype.toString.call(functionToCheck);
+  return objectProto === '[object Function]' || objectProto === '[object AsyncFunction]';
 }
 
 internals.implementation = function (server, options) {
 
   var validationResult = Joi.validate(options, optionsSchema);
-
-  if (validationResult.error){
+  if (validationResult.error) {
     throw new Error(validationResult.error.message);
   }
 
   var settings = Hoek.clone(options);
 
   var scheme = {
-    authenticate: function (request, reply) {
+    authenticate: async function (request, h) {
 
       var req = request.raw.req;
       var authorization = req.headers.authorization;
       if (!authorization) {
-        return reply(Boom.unauthorized(null, 'Bearer'));
+        throw Boom.unauthorized(null, 'Bearer');
       }
 
       var parts = authorization.split(/\s+/);
 
       if (parts.length !== 2) {
-        return reply(Boom.badRequest('Bad HTTP authentication header format', 'Bearer'));
+        throw Boom.badRequest('Bad HTTP authentication header format', 'Bearer');
       }
 
       if (parts[0].toLowerCase() !== 'bearer') {
-        return reply(Boom.unauthorized(null, 'Bearer'));
+        throw Boom.unauthorized(null, 'Bearer');
       }
 
-      if(parts[1].split('.').length !== 3) {
-        return reply(Boom.badRequest('Bad HTTP authentication header format', 'Bearer'));
+      if (parts[1].split('.').length !== 3) {
+        throw Boom.badRequest('Bad HTTP authentication header format', 'Bearer');
       }
 
       var token = parts[1];
 
       var getKey = isFunction(settings.key) ?
         settings.key :
-        function(req, token, callback) { callback(null, settings.key); };
+        function (req, token) { return { key: settings.key }; };
 
-      getKey(request, token, function(err, key, extraInfo){
-        if (err) { return reply(Boom.wrap(err)); }
-        // handle err
-        jwt.verify(token, key, settings, function(err, decoded) {
-          
-          if(err) {
-            return reply(Boom.unauthorized( 'JSON Web Token validation failed: ' + err.message, 'Bearer'));
-          }
-
-          if (!settings.validateFunc) {
-            return reply.continue({ credentials: decoded });
-          }
-
-          settings.validateFunc(decoded, extraInfo, function (err, isValid, credentials) {
-
-            credentials = credentials || null;
+      let keyResult;
+      try {
+        keyResult = await getKey(request, token);
+      } catch (err) {
+        throw err;
```

Comment:
This is the same as not having the try catch at all, we are throwing the same exception and doing nothing before throwing.

Reply:
I see that the original code is doing Boom.wrap, any reason to remove it?